### ISSUE TITLE
bug: CapacitorHttp - No support for Server-Sent Events (EventSource API)

### ISSUE THREAD

**Issue by torte** [ionic-team/capacitor#6582]
### Capacitor Version

  @capacitor/cli: 4.7.0
  @capacitor/core: 4.7.0
  @capacitor/ios: 4.7.0
  @capacitor/android: 4.7.0

### Platform(s)

iOS and Android regardless of version

### Current Behavior

When using the web's `EventSource` API for use with Server-Sent Events, the CapacitorHttp plugin won't be used and instead it falls back to the web's native implementation. This is caused CORS issues when intending to use device's native HTTP handling (hence the usage of CapacitorHttp). 

### Expected Behavior

As with `fetch` and `XMLHttpRequest`, `EventSource` should also be patched and work through device native events.

### Additional Context

Likely related to other issues similar to this since the `Content-Type` is `text/event-stream` for Server-Sent events: https://github.com/ionic-team/capacitor/issues/6299 & https://github.com/ionic-team/capacitor/pull/6206 ?

**Comment by jrobeson** [comment]
https://github.com/Azure/fetch-event-source lets you use SSE based on fetch(), but it still doesn't work with the capacitor http plugin because the plugin doesn't seem to deal with any kind of data stream at all.

**Comment by jrobeson** [comment]
I tried to use https://github.com/Azure/fetch-event-source to get a `fetch()` based approach for `text/event-stream`, but even that isn't good enough, since the patched fetch can't stream data.

**Comment by youngerjo** [comment]
I'd like to kindly request the Ionic team to consider adding support for the `text/event-stream` response type. This would greatly benefit scenarios like integrating OpenAI's chat completion with streaming, which is currently not achievable using the Capacitor's HTTP module. Thank you for considering this!

**Comment by jrobeson** [comment]
@youngerjo the type is not enough. the entire http client needs to be restructured to support streaming data.  I wonder if they should adopt OkHttp instead of trying to write it themselves

**Comment by youngerjo** [comment]
@jrobeson Thanks for pointing that out. For the time being, I might lean towards using the web fetch method for streaming responses. My main concern with that approach is the limitation of not being able to include an authorization header due to CORS restrictions. I genuinely hope the team considers restructuring the http client or integrating a more efficient solution like OkHttp in the future.

**Comment by torte** [comment]
Yeah, CORS has also been the main problem with our app just using native fetch. For just HTTP with simple content-types like `application/json` or `text/html` the current http plugin works fine, but everything else it does not. For now can not use SSE at all in the mobile app OR would need to make my app insecure essentially.

**Comment by maelp** [comment]
Is there also a way to use streaming HTTP request? I can do it from my web-app using `fetch` with a keep-alive connection, but it doesn't work once compiled to Capacitor

**Comment by IsaSalemHame** [comment]
Any update?

**Comment by jaspart** [comment]
Any update?

**Comment by th1nkgr33n** [comment]
Any Update?

**Comment by klauss194** [comment]
such a pitty that capacitor team wont reply to all these requests.
Guys I highly recommend if possible -> just move to a pwa. apple its not killing them anymore

**Comment by pizzaeater** [comment]
I'm missing support of SSE in the CapacitorHttp. Could we get any updates regarding this feature, please? Thank you.

**Comment by mdledoux** [comment]
What is the reason for even needing support in CapacitorHttp?  I actually use Angular and use @angular/http (HttpClient) in all of my services - so I haven't really needed the Capacitor plugin, which only a couple years ago was still a community plugin.

I have an assignment to implement SSE in my Angular app, and my plan is to write a service that directly uses the EventSource API - is there something wrong with this, or is this more of an issue for other frameworks which don't have a built-in HTTP client?

**Comment by mdledoux** [comment]
I'll also out that since this issue was opened in May of 2023 (using Cap 4.7.0), Capacitor 5 came and went, and Capacitor 6 came out April 15, 2024, and Cap 6.1 came out June 11, 2024.  I'd be curious if anyone has better luck with these in the latest version.

**Comment by torte** [comment]
@mdledoux You are right, you can achieve the same with regular web http client libraries and the `EventSource` API, but with caveats. One main caveat for me was that (at least at that time when I opened the ticket) you had to apply an open CORS policy on your backend, because the web API's run from the Capacitor mobile app don't set a trustworthy origin header (`localhost` isn't trustworthy on a production environment). There seems to be an option to overwrite that, but that didn't work for me, so I ended up trying `CapacitorHttp` and ran into more fundamental issues there (hence the ticket). 

Maybe the soluition is to be actually able to configure the `Origin` header instead of "falling" back to the native HTTP clients as `CapacitorHttp` does. That being said, setting an arbitary `Origin` header should also not be allowed in your config, because you would be able to hack other sites like this. So I guess the whole thing is a bit more complicated and there would be at least some validation process needed to achieve this (kind of how you validate your website with Android/iOS deep linking).

**Comment by hahagu** [comment]
https://stackoverflow.com/questions/59319882/capacitor-set-origin-for-cors

If you set a origin on the server, it will work.
Thanks for the hint, @torte

**Comment by untiny** [comment]
I encountered the same problem when integrating AI chat, which seems to be solvable through a worker approach, but compatibility was not considered.
``` javascript
function fetchWorker() {
  const workerMap = new Map();

  self.addEventListener("message", async (event) => {
    const { input, init, id, type } = event.data;
    if (type === "abort") {
      workerMap.get(id)?.cancel();
      workerMap.delete(id);
      return;
    }

    const controller = new AbortController();
    init.signal = controller.signal;
    try {
      const response = await fetch(input, init);

      if (response.status !== 200) {
        const data = await response.json();
        self.postMessage({
          id,
          error: data?.message || response.statusText,
        });
        return;
      }

      const reader = response.body?.getReader();

      if (!reader) {
        self.postMessage({ id, error: "Response body is not readable" });
        return;
      }

      workerMap.set(id, reader);

      while (true) {
        const { done, value } = await reader.read();
        if (done) {
          self.postMessage({ id, done: true });
          break;
        }
        self.postMessage({ id, chunk: value });
      }
    } catch (error: any) {
      console.error("Fetch error:", error);
      self.postMessage({ id, error: error.message });
    } finally {
      workerMap.delete(id);
    }
  });
}

function workerFetch(input: RequestInfo | URL, init?: RequestInit) {
  const blob = new Blob([`(${fetchWorker.toString()})()`], {
    type: "application/javascript",
  });
  const worker = new Worker(URL.createObjectURL(blob));

  const signal = init?.signal;

  const stream = new ReadableStream<Uint8Array>({
    start(controller) {
      const id = generateId();

      const onMessage = (
        event: MessageEvent<{
          id: string;
          chunk?: Uint8Array;
          done?: boolean;
          error?: string;
        }>
      ) => {
        const data = event.data;
        if (data.id !== id) return;

        if (data.chunk) {
          controller.enqueue(data.chunk);
        } else if (data.done) {
          controller.close();
          worker.terminate();
        } else if (data.error) {
          controller.error(new Error(data.error));
          worker.terminate();
        }
      };

      worker.addEventListener("message", onMessage);

      worker.postMessage({
        id,
        input,
        init: { ...init, signal: null },
      });

      if (signal) {
        const onAbort = () => {
          worker.postMessage({ id, type: "abort" });
          controller.error(new DOMException("Aborted", "AbortError"));
          worker.removeEventListener("message", onMessage);
          worker.terminate();
          signal.removeEventListener("abort", onAbort);
        };
        signal.addEventListener("abort", onAbort);
      }
    },
  });

  return Promise.resolve(
    new Response(stream, {
      status: 200,
      headers: { "Content-Type": "text/plain; charset=utf-8" },
    })
  );
}
```

**Comment by eric-horodyski** [comment]
We’re reviewing older issues as part of our backlog cleanup effort:  
[https://ionic.io/blog/keeping-capacitors-backlog-healthy](https://ionic.io/blog/keeping-capacitors-backlog-healthy)

To keep this open, please provide an updated minimal reproduction and environment details on the latest version of Capacitor within 7 days.

Otherwise, we’ll close it to keep the backlog focused. It can always be reopened with updated info.
